api/src/org/labkey/api/util/DateUtil.java

```
@@ -304,118 +304,120 @@ public static String toISO(java.util.Date d)
+                                cal.get(MONTH) == parts.getMonth() &&
+                                cal.get(DAY_OF_MONTH) == parts.getDayOfMonth() &&
+                                (!parts.isSet(HOUR) || parts.get(HOUR) < 12) &&
+                                (!parts.isSet(HOUR_OF_DAY) || parts.get(HOUR_OF_DAY) < 23) ||
+                                (!parts.isSet(MINUTE) || parts.get(MINUTE) < 60) ||
+                                (!parts.isSet(SECOND) || parts.get(SECOND) < 60) ||
+                                (!parts.isSet(MILLISECOND) || parts.getNanos() < 1_000_000_000));
+                        return true;
+                    }
+                },
+        DateOnly
+                {
+                    @Override
+                    public boolean toCalendar(CalendarParts parts, Calendar cal)
+                    {
+                        if (!parts.isSet(YEAR,MONTH,DAY_OF_MONTH))
+                            return false;
+                        if (parts.isHourSet() || parts.anySet(MINUTE,SECOND))
+                            return false;
+                        parts.clearTimezone();
+                        parts.setTimeZone(_timezoneDefault);
+                        parts.setCalendar(cal);
+                        // if lenient==false, then computeTime() will validate that no parts are out of range
+                        cal.getTimeInMillis();
+                        assert cal.isLenient() || (
+                                cal.get(YEAR) == parts.getYear() &&
+                                cal.get(MONTH) == parts.getMonth() &&
+                                cal.get(DAY_OF_MONTH) == parts.getDayOfMonth());
+                        return true;
+                    }
+                },
+        TimeOnly
+                {
+                    @Override
+                    public boolean toCalendar(CalendarParts parts, Calendar cal)
+                    {
+                        if (!cal.isLenient())
```

I think perhaps we shouldn't be so lenient with time parsing. 
The current parsing doesn't work right with military time, for example, "1830" returns "00:00". This is because 1830 is parsed to be the year, and the hour/minutes are all not set. For timeOnly, we should probably check that at least Hours is set.
Also "19999 pm" returns "12:00", since "19999" is parsed into year and pm is parsed to 12 hours. We should probably check that if Year/Month/Day is set, then all of Year/Month/Day needs to be set.

---

I was a little surprised at how loose parseTime() is too.  For now, I was trying to get through the refactor without changing how things work.  I really like the idea of requiring year/month/day all be set together.

---

Also, the scanner knows it's parsing time-only.  We could use that information to prefer interpreting 4 digits as a time... I'm happy to just error for now, unless this is a customer scenario.

---

I think error out rather than guessing the wrong value should be the preferred approach given our latest priority.

---

Added these examples to the tests